src/ems_client.ts

```diff
@@ -0,0 +1,474 @@
+  modifyUrlLocal(url, parsed => _.merge(parsed, props));
+
+/**
+ * plugins cannot have upstream dependencies on core/*-kibana.
+ * Work-around by copy-pasting modifyUrl routine here.
+ * @param url
+ * @param block
+ */
+function modifyUrlLocal(
+  url: string,
+  block: (urlParts: URLMeaningfulParts) => Partial<URLMeaningfulParts> | void
+): string {
+  const parsed = parseUrl(url, true) as URLMeaningfulParts;
+
+  // copy over the most specific version of each
+  // property. By default, the parsed url includes
+  // several conflicting properties (like path and
+  // pathname + search, or search and query) and keeping
+  // track of which property is actually used when they
+  // are formatted is harder than necessary
+  const meaningfulParts: URLMeaningfulParts = {
+    protocol: parsed.protocol,
+    slashes: parsed.slashes,
+    auth: parsed.auth,
+    hostname: parsed.hostname,
+    port: parsed.port,
+    pathname: parsed.pathname,
+    query: parsed.query || {},
+    hash: parsed.hash,
+  };
+
+  // the block modifies the meaningfulParts object, or returns a new one
+  const modifiedParts = block(meaningfulParts) || meaningfulParts;
+
+  // format the modified/replaced meaningfulParts back into a url
+  return formatUrl({
+    protocol: modifiedParts.protocol,
+    slashes: modifiedParts.slashes,
+    auth: modifiedParts.auth,
+    hostname: modifiedParts.hostname,
+    port: modifiedParts.port,
+    pathname: modifiedParts.pathname,
+    query: modifiedParts.query,
+    hash: modifiedParts.hash,
+  } as UrlObject);
+}
+
+/**
+ *  Unescape a url template that was escaped by encodeURI() so leaflet
+ *  will be able to correctly locate the variables in the template
+ *  @param  {String} url
+ *  @return {String}
+ */
+const unescapeTemplateVars = (url: string) => {
+  const ENCODED_TEMPLATE_VARS_RE = /%7B(\w+?)%7D/g;
+  return url.replace(ENCODED_TEMPLATE_VARS_RE, (total, varName) => `{${varName}}`);
+};
+
+type LocalizedStrings = { [key: string]: string };
+
+interface BaseClientConfig {
+  appName?: string;
+  manifestServiceUrl: string;
+  tileApiUrl: string;
+  fileApiUrl: string;
+  emsVersion?: string;
+  htmlSanitizer?: Function;
+  language?: string;
+  landingPageUrl?: string;
+  fetchFunction: Function;
+  proxyPath?: string;
+}
+
+interface DeprecatedClientConfig extends BaseClientConfig {
+  kbnVersion: string;
+}
+
+interface ClientConfig extends BaseClientConfig {
+  appVersion: string;
+}
+
+interface EmsCatalogService {
+  id?: string;
+  name?: string;
+  manifest: string;
+  type: string;
+}
+
+interface EmsCatalogManifest {
+  version?: string;
+  services: EmsCatalogService[];
+}
+
+interface EmsTmsCatalog {
+  version?: string;
+  services: ITMSService[];
+}
+
+export interface ITMSService {
+  id: string;
+  name: {
+    en: string;
+  };
+  attribution: EmsLayerAttribution[];
+  formats: {
+    locale: string;
+    format: string;
+    url: string;
+  }[];
```

Comment:
should this be extracted to its own type? I see it referenced implicitly several times later in the `TMSService` class